In [1]:
import pandas as pd
from datetime import datetime, timedelta
import io

In [2]:
users_df = pd.read_json('users.json',lines=True)
receipts_df = pd.read_json('receipts.json',lines=True)
brands_df = pd.read_json('brands.json',lines=True)

## Metadata
<br/>
Taking a look at the datasets on a high level to get an idea on some of the characteristics of the data we will be working with

### Receipts

In [3]:
#Total Number of records in the dataframe
len(receipts_df)

1119

In [4]:
# Column descriptions
# Non Null Counts of Each of the columns are visible.
# data type of each of the columns are mentioned
receipts_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1119 entries, 0 to 1118
Data columns (total 15 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   _id                      1119 non-null   object 
 1   bonusPointsEarned        544 non-null    float64
 2   bonusPointsEarnedReason  544 non-null    object 
 3   createDate               1119 non-null   object 
 4   dateScanned              1119 non-null   object 
 5   finishedDate             568 non-null    object 
 6   modifyDate               1119 non-null   object 
 7   pointsAwardedDate        537 non-null    object 
 8   pointsEarned             609 non-null    float64
 9   purchaseDate             671 non-null    object 
 10  purchasedItemCount       635 non-null    float64
 11  rewardsReceiptItemList   679 non-null    object 
 12  rewardsReceiptStatus     1119 non-null   object 
 13  totalSpent               684 non-null    float64
 14  userId                  

In [5]:
#Values are nested in columns like _id, createDate, dateScanned etc.,
receipts_df.describe(include='object')

_id     bonusPointsEarnedReason  \
count                                   1119                         544   
unique                                  1119                           9   
top     {'$oid': '5ff1e1eb0a720f0523000575'}  All-receipts receipt bonus   
freq                                       1                         183   

                      createDate               dateScanned  \
count                       1119                      1119   
unique                      1107                      1107   
top     {'$date': 1609687531000}  {'$date': 1609687531000}   
freq                           3                         3   

                    finishedDate                modifyDate  \
count                        568                      1119   
unique                       553                      1104   
top     {'$date': 1609687534000}  {'$date': 1609687536000}   
freq                           3                         3   

               pointsAwardedDate              purchaseDate  \
count                        537                       671   
unique                       523                       358   
top     {'$date': 1609687534000}  {'$date': 1597622400000}   
freq                           3                        40   

                                   rewardsReceiptItemList  \
count                                                 679   
unique                                                383   
top     [{'description': 'flipbelt level terrain waist...   
freq                                                   50   

       rewardsReceiptStatus                    userId  
count                  1119                      1119  
unique                    5                       258  
top                FINISHED  5fc961c3b8cfca11a077dd33  
freq                    518                       436

In [6]:
receipts_df.head()

_id  bonusPointsEarned  \
0  {'$oid': '5ff1e1eb0a720f0523000575'}              500.0   
1  {'$oid': '5ff1e1bb0a720f052300056b'}              150.0   
2  {'$oid': '5ff1e1f10a720f052300057a'}                5.0   
3  {'$oid': '5ff1e1ee0a7214ada100056f'}                5.0   
4  {'$oid': '5ff1e1d20a7214ada1000561'}                5.0   

                             bonusPointsEarnedReason  \
0  Receipt number 2 completed, bonus point schedu...   
1  Receipt number 5 completed, bonus point schedu...   
2                         All-receipts receipt bonus   
3                         All-receipts receipt bonus   
4                         All-receipts receipt bonus   

                 createDate               dateScanned  \
0  {'$date': 1609687531000}  {'$date': 1609687531000}   
1  {'$date': 1609687483000}  {'$date': 1609687483000}   
2  {'$date': 1609687537000}  {'$date': 1609687537000}   
3  {'$date': 1609687534000}  {'$date': 1609687534000}   
4  {'$date': 1609687506000}  {'$date': 1609687506000}   

               finishedDate                modifyDate  \
0  {'$date': 1609687531000}  {'$date': 1609687536000}   
1  {'$date': 1609687483000}  {'$date': 1609687488000}   
2                       NaN  {'$date': 1609687542000}   
3  {'$date': 1609687534000}  {'$date': 1609687539000}   
4  {'$date': 1609687511000}  {'$date': 1609687511000}   

          pointsAwardedDate  pointsEarned              purchaseDate  \
0  {'$date': 1609687531000}         500.0  {'$date': 1609632000000}   
1  {'$date': 1609687483000}         150.0  {'$date': 1609601083000}   
2                       NaN           5.0  {'$date': 1609632000000}   
3  {'$date': 1609687534000}           5.0  {'$date': 1609632000000}   
4  {'$date': 1609687506000}           5.0  {'$date': 1609601106000}   

   purchasedItemCount                             rewardsReceiptItemList  \
0                 5.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
1                 2.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
2                 1.0  [{'needsFetchReview': False, 'partnerItemId': ...   
3                 4.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
4                 2.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   

  rewardsReceiptStatus  totalSpent                    userId  
0             FINISHED        26.0  5ff1e1eacfcf6c399c274ae6  
1             FINISHED        11.0  5ff1e194b6a9d73a3a9f1052  
2             REJECTED        10.0  5ff1e1f1cfcf6c399c274b0b  
3             FINISHED        28.0  5ff1e1eacfcf6c399c274ae6  
4             FINISHED         1.0  5ff1e194b6a9d73a3a9f1052

### Brands

In [7]:
#Total Number of records in the dataframe
len(brands_df)

1167

In [8]:
# Column descriptions
# Non Null Counts of Each of the columns are visible.
# data type of each of the columns are mentioned
brands_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1167 entries, 0 to 1166
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   _id           1167 non-null   object 
 1   barcode       1167 non-null   int64  
 2   category      1012 non-null   object 
 3   categoryCode  517 non-null    object 
 4   cpg           1167 non-null   object 
 5   name          1167 non-null   object 
 6   topBrand      555 non-null    float64
 7   brandCode     933 non-null    object 
dtypes: float64(1), int64(1), object(6)
memory usage: 73.1+ KB


In [9]:
brands_df.describe(include='object')

_id category categoryCode  \
count                                   1167     1012          517   
unique                                  1167       23           14   
top     {'$oid': '601ac115be37ce2ead437551'}   Baking       BAKING   
freq                                       1      369          359   

                                                      cpg     name brandCode  
count                                                1167     1167       933  
unique                                                204     1156       897  
top     {'$ref': 'Cogs', '$id': {'$oid': '559c2234e4b0...  Huggies            
freq                                                   98        2        35

In [10]:
brands_df.head()

_id       barcode        category  \
0  {'$oid': '601ac115be37ce2ead437551'}  511111019862          Baking   
1  {'$oid': '601c5460be37ce2ead43755f'}  511111519928       Beverages   
2  {'$oid': '601ac142be37ce2ead43755d'}  511111819905          Baking   
3  {'$oid': '601ac142be37ce2ead43755a'}  511111519874          Baking   
4  {'$oid': '601ac142be37ce2ead43755e'}  511111319917  Candy & Sweets   

       categoryCode                                                cpg  \
0            BAKING  {'$id': {'$oid': '601ac114be37ce2ead437550'}, ...   
1         BEVERAGES  {'$id': {'$oid': '5332f5fbe4b03c9a25efd0ba'}, ...   
2            BAKING  {'$id': {'$oid': '601ac142be37ce2ead437559'}, ...   
3            BAKING  {'$id': {'$oid': '601ac142be37ce2ead437559'}, ...   
4  CANDY_AND_SWEETS  {'$id': {'$oid': '5332fa12e4b03c9a25efd1e7'}, ...   

                        name  topBrand                      brandCode  
0  test brand @1612366101024       0.0                            NaN  
1                  Starbucks       0.0                      STARBUCKS  
2  test brand @1612366146176       0.0  TEST BRANDCODE @1612366146176  
3  test brand @1612366146051       0.0  TEST BRANDCODE @1612366146051  
4  test brand @1612366146827       0.0  TEST BRANDCODE @1612366146827

### Receipts

In [11]:
len(users_df)

495

In [12]:
# Column descriptions
# Non Null Counts of Each of the columns are visible.
# data type of each of the columns are mentioned
users_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 495 entries, 0 to 494
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   _id           495 non-null    object
 1   active        495 non-null    bool  
 2   createdDate   495 non-null    object
 3   lastLogin     433 non-null    object
 4   role          495 non-null    object
 5   signUpSource  447 non-null    object
 6   state         439 non-null    object
dtypes: bool(1), object(6)
memory usage: 23.8+ KB


In [13]:
users_df.describe(include='object')

_id               createdDate  \
count                                    495                       495   
unique                                   212                       212   
top     {'$oid': '54943462e4b07e684157a532'}  {'$date': 1418998882381}   
freq                                      20                        20   

                       lastLogin      role signUpSource state  
count                        433       495          447   439  
unique                       172         2            2     8  
top     {'$date': 1614963143204}  consumer        Email    WI  
freq                          20       413          443   396

In [14]:
users_df.head()

_id  active               createdDate  \
0  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
1  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
2  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
3  {'$oid': '5ff1e1eacfcf6c399c274ae6'}    True  {'$date': 1609687530554}   
4  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   

                  lastLogin      role signUpSource state  
0  {'$date': 1609687537858}  consumer        Email    WI  
1  {'$date': 1609687537858}  consumer        Email    WI  
2  {'$date': 1609687537858}  consumer        Email    WI  
3  {'$date': 1609687530597}  consumer        Email    WI  
4  {'$date': 1609687537858}  consumer        Email    WI

## Data Transformation & Cleaning

In [15]:
#Function to treat values which are within a dictionary

from datetime import datetime


def treatment(cell):
  if isinstance(cell, dict):
      if '$oid' in cell:  
          return cell['$oid']
      elif '$date' in cell:
          # Convert the timestamp (milliseconds) to a datetime string
          return datetime.utcfromtimestamp(cell['$date'] / 1000).strftime('%Y-%m-%d %H:%M:%S')
  return cell

def series_treatment(row):
    if isinstance(row, dict):
        return pd.Series({'cpg_id': row.get('$id', None), 'ref': row.get('$ref', None)})
    else:
        return pd.Series({'id': None, 'ref': None})

def extract_data(df):
  return df.applymap(treatment)



def null_values_treatment(df):
    # Fill NaN values for string columns with 'unknown'
    string_columns = df.select_dtypes(include=['string']).columns
    df[string_columns] = df[string_columns].fillna('unknown')

    # Fill NaN values for numeric columns (float and int) with 0
    numeric_columns = df.select_dtypes(include=['float64', 'int64']).columns
    df[numeric_columns] = df[numeric_columns].fillna(0)

    # Return the updated DataFrame
    return df

In [16]:
#extract the values of cpg into 2 separate columns in the brand dataframe
cpg_df = pd.DataFrame(brands_df['cpg'])
brands_df[['cpg_id', 'ref']] = cpg_df['cpg'].apply(series_treatment)

#Extract all the data with $oid, $date etc.
users = extract_data(users_df)
brands = extract_data(brands_df)
receipts = extract_data(receipts_df)

In [17]:
# Renaming the _id columns to clearly differentiate between IDs in the dataframes

users = users.rename(columns={'_id': 'userId'})
brands = brands.rename(columns={'_id': 'brandId','name': 'brandName'})
receipts = receipts.rename(columns={'_id': 'receiptId'})

### Creating ReceiptsItemList

<br/> Extracting the 'receiptItemslist' from the receipts dataframe in order to create a new table with its own unique UUID and holding the data about the item lists along with the receiptID as the foreign key

In [18]:
#creating a new receiptsItemList dataframe that will unpack the data from rewardsReceiptItemList
import uuid

def createReceiptItemList(df):
  unpacked_data = []
  for index, row in df.iterrows():
    receiptID = row['receiptId']
    items = row['rewardsReceiptItemList']
    if isinstance(items, list):
      for item in items:
        partnerItemId = item['partnerItemId']
        new_uuid = str(uuid.uuid5(uuid.NAMESPACE_DNS,f"{receiptID}-{partnerItemId}"))
        unpacked_data.append({'receiptsItemListId':new_uuid,'receiptId': receiptID, **item})
    # else:
    #   print(f'Warning: Row {index} does not have valid Items List \n {items} \n')
  return pd.DataFrame(unpacked_data)

receiptsItemList_df = createReceiptItemList(receipts)

len(receiptsItemList_df)


6941

## Data Quality Issues

### Duplicate Checks

These are basic duplicates checks performed on Primary keys of the table and other custom columns based on business needs to ensure data looks good. <br/>

In [19]:
def check_duplicates(df, primary_key=None, duplicate_columns=None):
    results = []
    
    # Check duplicates on primary key
    if primary_key:
        pk_duplicates = df.duplicated(subset=[primary_key]).sum()
        results.append({'Check': 'Primary Key Duplicates', 'Column(s)': primary_key, 'Count': pk_duplicates})
    
    # Check duplicates on specific columns
    if duplicate_columns:
        col_duplicates = df.duplicated(subset=duplicate_columns).sum()
        results.append({'Check': 'Duplicate Columns', 'Column(s)': ', '.join(duplicate_columns), 'Count': col_duplicates})
    
    return pd.DataFrame(results)

In [20]:
users_column_list = [col for col in users.columns if col not in 'userId']
users_duplicate_checks= check_duplicates(users, 'userId',users_column_list)
#ideally duplicate checks are generally performed mainly on primary keys. 
# I have added the ability to include multiple column checks as well based on customisable needs on dataframes
users_duplicate_checks

Check                                          Column(s)  \
0  Primary Key Duplicates                                             userId   
1       Duplicate Columns  active, createdDate, lastLogin, role, signUpSo...   

   Count  
0    283  
1    283

In [21]:
#Sample to show duplicates existing in Users
users.groupby('userId')[['active']].count().sort_values(by=['active'],ascending=False)

active
userId                          
54943462e4b07e684157a532      20
5fc961c3b8cfca11a077dd33      20
5fa41775898c7a11a6bcef3e      18
59c124bae4b0299e55b0f330      18
5ff5d15aeb7c7d12096d91a2      18
...                          ...
5ff3711e62fde912123a620e       1
60074883325c8a1794622a12       1
60074b996e64691717e8f11a       1
60083a1e325c8a17946255de       1
60268c7aefa6011bb1510786       1

[212 rows x 1 columns]

In [22]:
#Duplicates observed on a particular user ID as follows
users[users['userId']=='5ff1e194b6a9d73a3a9f1052']

userId  active          createdDate  \
0   5ff1e194b6a9d73a3a9f1052    True  2021-01-03 15:24:04   
1   5ff1e194b6a9d73a3a9f1052    True  2021-01-03 15:24:04   
2   5ff1e194b6a9d73a3a9f1052    True  2021-01-03 15:24:04   
4   5ff1e194b6a9d73a3a9f1052    True  2021-01-03 15:24:04   
5   5ff1e194b6a9d73a3a9f1052    True  2021-01-03 15:24:04   
8   5ff1e194b6a9d73a3a9f1052    True  2021-01-03 15:24:04   
10  5ff1e194b6a9d73a3a9f1052    True  2021-01-03 15:24:04   
12  5ff1e194b6a9d73a3a9f1052    True  2021-01-03 15:24:04   
14  5ff1e194b6a9d73a3a9f1052    True  2021-01-03 15:24:04   
15  5ff1e194b6a9d73a3a9f1052    True  2021-01-03 15:24:04   
19  5ff1e194b6a9d73a3a9f1052    True  2021-01-03 15:24:04   

              lastLogin      role signUpSource state  
0   2021-01-03 15:25:37  consumer        Email    WI  
1   2021-01-03 15:25:37  consumer        Email    WI  
2   2021-01-03 15:25:37  consumer        Email    WI  
4   2021-01-03 15:25:37  consumer        Email    WI  
5   2021-01-03 15:25:37  consumer        Email    WI  
8   2021-01-03 15:25:37  consumer        Email    WI  
10  2021-01-03 15:25:37  consumer        Email    WI  
12  2021-01-03 15:25:37  consumer        Email    WI  
14  2021-01-03 15:25:37  consumer        Email    WI  
15  2021-01-03 15:25:37  consumer        Email    WI  
19  2021-01-03 15:25:37  consumer        Email    WI

In [23]:
brands_duplicate_checks= check_duplicates(brands, 'brandId')
brands_duplicate_checks

Check Column(s)  Count
0  Primary Key Duplicates   brandId      0

In [24]:
receipts_duplicate_checks = check_duplicates(receipts, 'receiptId')
receipts_duplicate_checks

Check  Column(s)  Count
0  Primary Key Duplicates  receiptId      0

About 283 records for userID (primary key) were observed having duplicates for <b>Users</b>.
<br/> <b>brands, and receipts</b> did not have any duplicates observed on the primary keys

### Non Null Checks

These are custom checks to take a look at the dataframe to see what values of the dataframe have null values in them and treat them in the following section accordingly in order to prevent errors during data transformations further

In [25]:
def check_non_nulls(df):
    null_counts = df.isnull().sum()
    null_percentages = (df.isnull().mean() * 100).round(2)
    
    report = pd.DataFrame({
        'Column': null_counts.index,
        'Null Count': null_counts.values,
        'Null Percentage': null_percentages.values
    })
    
    return report

In [26]:
users_null_check = check_non_nulls(users_df)
users_null_check

Column  Null Count  Null Percentage
0           _id           0             0.00
1        active           0             0.00
2   createdDate           0             0.00
3     lastLogin          62            12.53
4          role           0             0.00
5  signUpSource          48             9.70
6         state          56            11.31

In [27]:
brands_null_check = check_non_nulls(brands_df)
brands_null_check

Column  Null Count  Null Percentage
0           _id           0             0.00
1       barcode           0             0.00
2      category         155            13.28
3  categoryCode         650            55.70
4           cpg           0             0.00
5          name           0             0.00
6      topBrand         612            52.44
7     brandCode         234            20.05
8        cpg_id           0             0.00
9           ref           0             0.00

In [28]:
receipts_null_check = check_non_nulls(receipts_df)
receipts_null_check

Column  Null Count  Null Percentage
0                       _id           0             0.00
1         bonusPointsEarned         575            51.39
2   bonusPointsEarnedReason         575            51.39
3                createDate           0             0.00
4               dateScanned           0             0.00
5              finishedDate         551            49.24
6                modifyDate           0             0.00
7         pointsAwardedDate         582            52.01
8              pointsEarned         510            45.58
9              purchaseDate         448            40.04
10       purchasedItemCount         484            43.25
11   rewardsReceiptItemList         440            39.32
12     rewardsReceiptStatus           0             0.00
13               totalSpent         435            38.87
14                   userId           0             0.00

For the null checks, I have performed a null value treatment converting all the strings to 'unknown' and the floats and integers to 0. While preserving the boolean values later on.

### Data Recency checks

These checks are usually performed on fields like createdDate and modifiedDate to keep a recency on the data because on data model and pipelines one of the biggest issues could arise out of stale data leading to wrong reports

In [30]:
def data_recency_check(df, date_columns):
    threshold_date = pd.to_datetime(datetime.now().date() - timedelta(days=1))
    results = []
    for column in date_columns:
        col_val = df[df[column].notnull()][[column]]
        max_date = pd.to_datetime(col_val[column].max())
        is_valid = max_date >= threshold_date if pd.notnull(max_date) else False
        results.append({
                    'Column': column,
                    'Max Date': max_date,
                    'Threshold Date': threshold_date,
                    'Valid': is_valid
                })
    report = pd.DataFrame(results)
    
    return report

In [31]:
user_report = data_recency_check(users, ['createdDate','lastLogin'])

In [32]:
user_report

Column            Max Date Threshold Date  Valid
0  createdDate 2021-02-12 14:11:06     2025-01-22  False
1    lastLogin 2021-03-05 16:52:23     2025-01-22  False

In [33]:
receipts_report = data_recency_check(receipts, ['createDate','dateScanned','finishedDate','modifyDate','pointsAwardedDate','purchaseDate'])

In [34]:
receipts_report

Column            Max Date Threshold Date  Valid
0         createDate 2021-03-01 23:17:34     2025-01-22  False
1        dateScanned 2021-03-01 23:17:34     2025-01-22  False
2       finishedDate 2021-02-26 22:36:25     2025-01-22  False
3         modifyDate 2021-03-01 23:17:34     2025-01-22  False
4  pointsAwardedDate 2021-02-26 22:36:25     2025-01-22  False
5       purchaseDate 2021-03-08 17:37:13     2025-01-22  False

### Data Quality Checks on ReceiptsItemList

In [35]:
receiptsItemList_duplicate_checks = check_duplicates(receiptsItemList_df, 'receiptsItemListId')
receiptsItemList_duplicate_checks

Check           Column(s)  Count
0  Primary Key Duplicates  receiptsItemListId      0

In [36]:
receiptsItemList_null_check = check_non_nulls(receiptsItemList_df)
receiptsItemList_null_check

Column  Null Count  Null Percentage
0                   receiptsItemListId           0             0.00
1                            receiptId           0             0.00
2                              barcode        3851            55.48
3                          description         381             5.49
4                           finalPrice         174             2.51
5                            itemPrice         174             2.51
6                     needsFetchReview        6128            88.29
7                        partnerItemId           0             0.00
8               preventTargetGapPoints        6583            94.84
9                    quantityPurchased         174             2.51
10                  userFlaggedBarcode        6604            95.14
11                  userFlaggedNewItem        6618            95.35
12                    userFlaggedPrice        6642            95.69
13                 userFlaggedQuantity        6642            95.69
14              needsFetchReviewReason        6722            96.84
15              pointsNotAwardedReason        6601            95.10
16                       pointsPayerId        5674            81.75
17                        rewardsGroup        5210            75.06
18             rewardsProductPartnerId        4672            67.31
19              userFlaggedDescription        6736            97.05
20            originalMetaBriteBarcode        6870            98.98
21        originalMetaBriteDescription        6931            99.86
22                           brandCode        4341            62.54
23              competitorRewardsGroup        6666            96.04
24                 discountedItemPrice        1172            16.89
25             originalReceiptItemText        1181            17.01
26                          itemNumber        6788            97.80
27  originalMetaBriteQuantityPurchased        6926            99.78
28                        pointsEarned        6014            86.64
29                         targetPrice        6563            94.55
30                  competitiveProduct        6296            90.71
31                  originalFinalPrice        6932            99.87
32          originalMetaBriteItemPrice        6932            99.87
33                             deleted        6932            99.87
34                    priceAfterCoupon        5985            86.23
35                 metabriteCampaignId        6078            87.57

## Datatype Setting

In [37]:
#Null Values Treatment

users = null_values_treatment(users)
brands = null_values_treatment(brands)
receipts = null_values_treatment(receipts)
receiptsItemList = null_values_treatment(receiptsItemList_df)

In [38]:
users = users.astype({'userId': 'string', 'active': 'bool', 'role':'string', 'signUpSource':'string', 'state':'string'})
# users['createdDate'] = pd.to_datetime(users['createdDate'])
# users['lastLogin'] = pd.to_datetime(users['lastLogin'])

users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 495 entries, 0 to 494
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   userId        495 non-null    string
 1   active        495 non-null    bool  
 2   createdDate   495 non-null    object
 3   lastLogin     433 non-null    object
 4   role          495 non-null    string
 5   signUpSource  447 non-null    string
 6   state         439 non-null    string
dtypes: bool(1), object(2), string(4)
memory usage: 23.8+ KB


In [39]:
brands = brands.astype({
    "brandId": "string",
    "barcode": "string",
    "category": "string",
    "categoryCode": "string",
    "cpg": "string",
    "brandName": "string",
    "topBrand": "float64",
    "brandCode": "string",
    "cpg_id": "string",
    "ref": "string"
})
brands.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1167 entries, 0 to 1166
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   brandId       1167 non-null   string 
 1   barcode       1167 non-null   string 
 2   category      1012 non-null   string 
 3   categoryCode  517 non-null    string 
 4   cpg           1167 non-null   string 
 5   brandName     1167 non-null   string 
 6   topBrand      1167 non-null   float64
 7   brandCode     933 non-null    string 
 8   cpg_id        1167 non-null   string 
 9   ref           1167 non-null   string 
dtypes: float64(1), string(9)
memory usage: 91.3 KB


In [40]:
receipts = receipts.astype({
    "receiptId": "string",
    "bonusPointsEarned": "float64",
    "bonusPointsEarnedReason": "string",
    "modifyDate": "datetime64[ns]",
    "pointsEarned": "float64",
    "purchasedItemCount": "float64",
    "rewardsReceiptItemList": "string",
    "rewardsReceiptStatus": "string",
    "totalSpent": "float64",
    "userId": "string"
})

receipts.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1119 entries, 0 to 1118
Data columns (total 15 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   receiptId                1119 non-null   string        
 1   bonusPointsEarned        1119 non-null   float64       
 2   bonusPointsEarnedReason  544 non-null    string        
 3   createDate               1119 non-null   object        
 4   dateScanned              1119 non-null   object        
 5   finishedDate             568 non-null    object        
 6   modifyDate               1119 non-null   datetime64[ns]
 7   pointsAwardedDate        537 non-null    object        
 8   pointsEarned             1119 non-null   float64       
 9   purchaseDate             671 non-null    object        
 10  purchasedItemCount       1119 non-null   float64       
 11  rewardsReceiptItemList   679 non-null    string        
 12  rewardsReceiptStatus     1119 non-

In [41]:
receiptsItemList = receiptsItemList.astype({
    "receiptsItemListId": "string",
    "receiptId": "string",
    "barcode": "string",
    "description": "string",
    "finalPrice": "float64",
    "itemPrice": "float64",
    "needsFetchReview": "bool",
    "partnerItemId": "int64",  
    "preventTargetGapPoints": "bool",
    "quantityPurchased": "float64",
    "userFlaggedBarcode": "string",  
    "userFlaggedNewItem": "bool",
    "userFlaggedPrice": "float64",
    "userFlaggedQuantity": "float64",
    "needsFetchReviewReason": "string",
    "pointsNotAwardedReason": "string",
    "pointsPayerId": "string",
    "rewardsGroup": "string",
    "rewardsProductPartnerId": "string",
    "userFlaggedDescription": "string",
    "originalMetaBriteBarcode": "string",
    "originalMetaBriteDescription": "string",
    "brandCode": "string",
    "competitorRewardsGroup": "string",
    "discountedItemPrice": "float64",
    "originalReceiptItemText": "string",
    "itemNumber": "string",  
    "originalMetaBriteQuantityPurchased": "float64",
    "pointsEarned": "float64",
    "targetPrice": "float64",
    "competitiveProduct": "bool",
    "originalFinalPrice": "float64",
    "originalMetaBriteItemPrice": "float64",
    "deleted": "bool",
    "priceAfterCoupon": "float64",
    "metabriteCampaignId": "string"
})

receiptsItemList.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6941 entries, 0 to 6940
Data columns (total 36 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   receiptsItemListId                  6941 non-null   string 
 1   receiptId                           6941 non-null   string 
 2   barcode                             3090 non-null   string 
 3   description                         6560 non-null   string 
 4   finalPrice                          6767 non-null   float64
 5   itemPrice                           6767 non-null   float64
 6   needsFetchReview                    6941 non-null   bool   
 7   partnerItemId                       6941 non-null   int64  
 8   preventTargetGapPoints              6941 non-null   bool   
 9   quantityPurchased                   6941 non-null   float64
 10  userFlaggedBarcode                  337 non-null    string 
 11  userFlaggedNewItem                  6941 no

In [42]:
receiptsItemList = null_values_treatment(receiptsItemList)

In [43]:
#Generating the CSV Files to be added over in DBeaver SQL (SQLite) for further querying and Analysis
users.to_csv('users_data.csv')
brands.to_csv('brands_data.csv')
receipts.to_csv('receipts_data.csv')
receiptsItemList.to_csv('receiptsItemList_data.csv')

In [44]:
receiptsItemList.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6941 entries, 0 to 6940
Data columns (total 36 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   receiptsItemListId                  6941 non-null   string 
 1   receiptId                           6941 non-null   string 
 2   barcode                             6941 non-null   string 
 3   description                         6941 non-null   string 
 4   finalPrice                          6941 non-null   float64
 5   itemPrice                           6941 non-null   float64
 6   needsFetchReview                    6941 non-null   bool   
 7   partnerItemId                       6941 non-null   int64  
 8   preventTargetGapPoints              6941 non-null   bool   
 9   quantityPurchased                   6941 non-null   float64
 10  userFlaggedBarcode                  6941 non-null   string 
 11  userFlaggedNewItem                  6941 no